cmd/zetaclientd/utils.go

```diff
@@ -97,9 +80,68 @@ func waitForZetacoreToCreateBlocks(ctx context.Context, zc interfaces.ZetacoreCl
 			return fmt.Errorf("zetacore is not ready, timeout %s", time.Since(start).String())
 		}
 
-		logger.Info().Msgf("Failed to get block number, retry : %d/%d", retryCount, attempts)
-		time.Sleep(interval)
+		logger.Info().Msgf("Failed to get block number, retry: %d/%d", retryCount, attempts)
+		return nil
+	}
+
+	return ticker.Run(ctx, interval, task)
+}
+
+// prepareZetacoreClient prepares the zetacore client for use.
+// EXITS THE PROGRAM IF THIS NODE IS NOT AN OBSERVER.
+func prepareZetacoreClient(ctx context.Context, zc *zetacore.Client, cfg *config.Config, logger zerolog.Logger) error {
+	// Set grantee account number and sequence number
+	if err := zc.SetAccountNumber(authz2.ZetaClientGranteeKey); err != nil {
+		return errors.Wrap(err, "failed to set account number")
+	}
+
+	res, err := zc.GetNodeInfo(ctx)
+	if err != nil {
+		return errors.Wrap(err, "failed to get node info")
+	}
+
+	network := res.GetDefaultNodeInfo().Network
+	if network != cfg.ChainID {
+		logger.Warn().
+			Str("got", cfg.ChainID).
+			Str("want", network).
+			Msg("Zetacore chain id config mismatch. Forcing update from the network")
+
+		cfg.ChainID = network
+		if err = zc.UpdateChainID(cfg.ChainID); err != nil {
+			return errors.Wrap(err, "failed to update chain id")
+		}
+	}
+
+	isObserver, err := isObserverNode(ctx, zc)
+	switch {
+	case err != nil:
+		return errors.Wrap(err, "failed to check if this node is an observer")
+	case !isObserver:
+		addr := zc.GetKeys().GetOperatorAddress().String()
+		logger.Info().Str("operator_address", addr).Msg("This node is not an observer. Exit 0")
+		os.Exit(0)
 	}
+
+	return nil
+}
+
+// isObserverNode checks whether THIS node is an observer node.
+func isObserverNode(ctx context.Context, zc *zetacore.Client) (bool, error) {
+	observers, err := zc.GetObserverList(ctx)
+	if err != nil {
+		return false, errors.Wrap(err, "unable to get observers list")
+	}
+
+	operatorAddress := zc.GetKeys().GetOperatorAddress().String()
+
+	for _, observer := range observers {
+		if observer == operatorAddress {
+			return true, nil
+		}
+	}
```

Comment:
Consider adding a function for observerSet, something like IsObserver(). 
We can put the function under x/observer/types.

Reply:
Same as here https://github.com/zeta-chain/node/pull/3291#discussion_r1883839054

Reply:
I realised that it's `QueryObserverSetResponse`, not `ObserverSet` that it used within the keeper. Will keep as is